### PR TITLE
Configurable Spreadsheet and Transaction Providers

### LINKED ISSUE #2: Pluggable interface for transaction data
As mentioned on [HN thread](https://news.ycombinator.com/item?id=19014820) – there are some security concerns about using Plaid.

As mentioned on [Reddit thread](https://www.reddit.com/r/javascript/comments/akidk9/mintable_a_rollyourown_mint_clone_for_managing/) – some users have banks which are not supported with Plaid.

### PR DESCRIPTION
**Pull request by kevinschaich**
Fixes #2 & #3 

### Transaction Provider

`TRANSACTION_PROVIDER` specifies which service to use to fetch transactions. At this time, the only possible value is `"plaid"`, but we plan to add other providers in the future.

### Spreadsheet Provider

`SPREADSHEET_PROVIDER` specifies which service to use to automate spreadsheet updates. At this time, the only possible value is `"sheets"`, but we plan to add other providers in the future.

### CODE REVIEW

**Comment by kevinschaich** [comment]
@johnwiseheart

File: index.js [hunk 1]
```diff
@@ -42,114 +29,40 @@ const {
...
-      if (new RegExp(override.pattern, _.get(override, 'flags', '')).test(transaction.name)) {
-        transaction['category.0'] = _.get(override, 'category.0', '');
-        transaction['category.1'] = _.get(override, 'category.1', '');
-      }
-    });
-
-    return _.at(transaction, transactionColumns)
+  switch (spreadsheetProvider) {
+    case 'sheets':
+      await updateSheets(
```
Comment (kevinschaich) [inline index.js]: I'm not sure if there's a better way to do this in Node – looks kind of ugly with all these constants passed through like this.

I think these would be things we'd want defined in `index.js` to pass to other providers, though.

File: index.js [hunk 2]
```diff
@@ -1,32 +1,19 @@
...
-
   const defaultTransactionColumns = ['date', 'amount', 'name', 'account', 'category.0', 'category.1', 'pending'];
   const defaultReferenceColumns = ['notes', 'work', 'joint'];
+  const defaultSpreadsheetProvider = 'sheets';
+  const defaultTransactionProvider = 'plaid';
+
   const transactionColumns = JSON.parse(process.env.TRANSACTION_COLUMNS || null) || defaultTransactionColumns;
   const referenceColumns = JSON.parse(process.env.REFERENCE_COLUMNS || null) || defaultReferenceColumns;
   const categoryOverrides = JSON.parse(process.env.CATEGORY_OVERRIDES || null) || [];
+  const spreadsheetProvider = JSON.parse(process.env.SPREADSHEET_PROVIDER || null) || defaultSpreadsheetProvider;
```
Comment (johnwiseheart) [inline index.js]: Probably worth extracting this into a function `getEnvOrDefault(process.env.SPREADSHEET_PROVIDER, defaultSpreadsheetProvider)` or similar

File: index.js (hunk 1 above)
Comment (johnwiseheart) [reply index.js]: I think its fine - ideally when we add additional providers the interface should be the same - that way we don't have to pass everything in for every additional provider, we can just have one call where we can change the method.

File: providers/plaid.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
+const moment = require('moment');
```
Comment (johnwiseheart) [inline providers/plaid.js]: most of this code has just moved I think

File: providers/sheets.js [hunk 4]
```diff
@@ -0,0 +1,78 @@
+const _ = require('lodash');
```
Comment (johnwiseheart) [inline providers/sheets.js]: same here

**Comment by kevinschaich** [comment]
@johnwiseheart updated according to your comments. I'll open a new issue for the "interface" bit.

**Pull request merged**